clients/drcachesim/common/lz4_istream.h

```diff
@@ -0,0 +1,146 @@
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* lz4_istream_t: a wrapper around lz4 to match the parts of the
+ * std::istream interface we use for raw2trace and file_reader_t.
+ * Supports only limited seeking within the current internal buffer.
+ */
+
+#ifndef _LZ4_ISTREAM_H_
+#define _LZ4_ISTREAM_H_ 1
+
+#ifndef HAS_LZ4
+#    error HAS_LZ4 is required
+#endif
+#include <fstream>
+#include <lz4frame.h>
+#include <iostream>
+
+/* We need to override the stream buffer class which is where the file
+ * reads happen.  The stream buffer base class reads from eback()..egptr()
+ * with the next to read at gptr().
+ */
+class lz4_istreambuf_t : public std::basic_streambuf<char, std::char_traits<char>> {
+public:
+    lz4_istreambuf_t(const std::string &path)
+    {
+        if (buffer_size_ <= LZ4F_HEADER_SIZE_MAX)
+            return;
+        file_ = fopen(path.c_str(), "rb");
+        if (file_ == nullptr)
+            return;
+        size_t res = LZ4F_createDecompressionContext(&lzcxt_, LZ4F_VERSION);
+        if (LZ4F_isError(res)) {
+            fclose(file_);
+            file_ = nullptr;
+            return;
+        }
+        buf_compressed_ = new char[buffer_size_];
+        buf_uncompressed_ = new char[buffer_size_];
+        cur_src_ = buf_compressed_;
+    }
+    ~lz4_istreambuf_t() override
+    {
+        if (file_ != nullptr) {
+            fclose(file_);
+        }
+        delete[] buf_compressed_;
+        delete[] buf_uncompressed_;
+        LZ4F_freeDecompressionContext(lzcxt_);
+    }
+    int
+    underflow() override
+    {
+        if (file_ == nullptr)
+            return traits_type::eof();
+        if (gptr() == egptr()) {
+            size_t dst_sz;
+            do {
+                if (src_left_ == 0) {
+                    src_left_ = fread(buf_compressed_, 1, buffer_size_, file_);
+                    if (ferror(file_) || src_left_ == 0)
+                        return traits_type::eof();
+                    cur_src_ = buf_compressed_;
+                }
+                size_t src_sz = src_left_;
+                dst_sz = buffer_size_;
+                size_t res = LZ4F_decompress(lzcxt_, buf_uncompressed_, &dst_sz, cur_src_,
+                                             &src_sz, nullptr);
+                if (LZ4F_isError(res))
+                    return traits_type::eof();
+                src_left_ -= src_sz;
+                cur_src_ += src_sz;
+            } while (dst_sz == 0);
+            setg(buf_uncompressed_, buf_uncompressed_, buf_uncompressed_ + dst_sz);
+        }
+        return *gptr();
+    }
+    std::iostream::pos_type
+    seekoff(std::iostream::off_type off, std::ios_base::seekdir dir,
+            std::ios_base::openmode which = std::ios_base::in) override
+    {
+        if (dir == std::ios_base::cur &&
+            ((off >= 0 && gptr() + off < egptr()) ||
+             (off < 0 && gptr() + off >= eback())))
+            gbump(off);
```

Comment:
Should this return 0 after the call to gbump since gbump is void?

Reply:
The return value is the current position, which will not be 0 (unless the caller seek-ed to 0): the gptr() - eback() looks correct since gbump should update gptr.